### PR TITLE
Python 3.9 migration

### PR DESCRIPTION
**Pull request by sherwinlu**
Migrating the project to be Python 3.9 compatible.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,24 +3,22 @@ workflows:
    version: 2
    eb-sqs-jobs:
      jobs:
-       - test-python-3-7
-       - test-python-2-7
+       - test-python-3-9
 jobs:
-  test-python-3-7: &test-template
+  test-python-3-9: &test-template
```
Comment (gurkansokat) [inline .circleci/config.yml:8]: I think we can remove the test-template.

File: .circleci/config.yml [hunk 2]
```diff
@@ -3,24 +3,22 @@ workflows:
...
       - run:
           name: Run tests
           command: |
-            python -m django test --settings=eb_sqs.test_settings
-  test-python-2-7:
-    <<: *test-template
-    docker:
-      - image: circleci/python:2.7.16
+            . venv/bin/activate
+            python -m django test --settings=eb_sqs.test_settings
```
Comment (gurkansokat) [inline .circleci/config.yml:24]: Nit. 
New line at the end of file.

File: eb_sqs/auto_tasks/service.py [hunk 3]
```diff
@@ -54,12 +55,12 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
                 exc.max_retries_func()
             else:
                 # by default log an error
-                logger.error('Reached max retries in auto task {}.{}.{} with error: {}'.format(module_name, class_name, func_name, repr(exc)))
+                logger.error('Reached max retries in auto task {}.{}.{} with error: {}'.format(module_name, class_name,
```
Comment (gurkansokat) [inline eb_sqs/auto_tasks/service.py:58]: Should we convert these to f-string way now?

File: eb_sqs/worker/worker_task.py [hunk 4]
```diff
@@ -76,13 +74,11 @@ def copy(self, use_serialization):
...
     @staticmethod
-    def _pickle_args(args):
-        # type: (Any) -> unicode
+    def _pickle_args(args: Any) -> str:
         return base64.b64encode(pickle.dumps(args, pickle.HIGHEST_PROTOCOL)).decode('utf-8')
 
     @staticmethod
-    def deserialize(msg):
-        # type: (unicode) -> WorkerTask
+    def deserialize(msg: str):
```
Comment (gurkansokat) [inline eb_sqs/worker/worker_task.py:81]: Need to add -> WorkerTask for return?

File: eb_sqs/auto_tasks/service.py (hunk 3 above)
Comment (gurkansokat) [reply eb_sqs/auto_tasks/service.py:58]: There are other places that we use the old .format way, but not sure if it is too much work or not to convert all to f-string. What do you think?

File: eb_sqs/worker/worker_task.py [hunk 5]
```diff
@@ -30,15 +30,13 @@ def __init__(self, id, group_id, queue, func, args, kwargs, max_retries, retry,
...
-        # type: () -> Any
+    def execute(self) -> Any:
         from eb_sqs.decorators import func_retry_decorator
         self.func.retry_num = self.retry
         self.func.retry = func_retry_decorator(worker_task=self)
         return self.func(*self.args, **self.kwargs)
 
-    def serialize(self):
-        # type: () -> unicode
+    def serialize(self) -> str:
```
Comment (gurkansokat) [inline eb_sqs/worker/worker_task.py:39]: Not related to this particular line, but below we have missed the syntax update for the "copy" function

File: eb_sqs/worker/worker.py [hunk 6]
```diff
@@ -69,19 +68,19 @@ def execute(self, msg):
 
             raise ExecutionFailedException(worker_task.abs_func_name, ex)
 
-    def delay(self, group_id, queue_name, func, args, kwargs, max_retries, use_pickle, delay, execute_inline):
-        # type: (unicode, unicode, Any, tuple, dict, int, bool, int, bool) -> Any
-        worker_task = WorkerTask(str(uuid.uuid4()), group_id, queue_name, func, args, kwargs, max_retries, 0, None, use_pickle)
+    def delay(self, group_id: str, queue_name: str, func, args: tuple, kwargs: dict, max_retries: int, use_pickle: bool,
```
Comment (gurkansokat) [inline eb_sqs/worker/worker.py:71]: func parameter missing type. (func: Any)

File: eb_sqs/worker/service.py [hunk 7]
```diff
@@ -4,8 +4,13 @@
 import signal
 from datetime import timedelta
 from time import sleep
+from typing import Any
 
 import boto3
+from boto.sqs.message import Message
```
Comment (gurkansokat) [inline eb_sqs/worker/service.py:10]: are these the right from packages? My PyCharm was complaining and I couldn't find an example of this anywhere.

File: eb_sqs/auto_tasks/service.py [hunk 8]
```diff
@@ -54,12 +55,12 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
...
+                                                                                               func_name, repr(exc)))
 
 
 class AutoTaskService(object):
-    def register_task(self, method, queue_name=None, max_retries=None):
-        # type: (Any, str, int) -> None
+    def register_task(self, method, queue_name: str = None, max_retries: int = None):
```
Comment (gurkansokat) [inline eb_sqs/auto_tasks/service.py:63]: "method -> Any" or we skip it on purpose?

File: eb_sqs/auto_tasks/exceptions.py [hunk 9]
```diff
@@ -1,12 +1,10 @@
 class RetryableTaskException(Exception):
-    def __init__(self, inner, delay=None, count_retries=None, max_retries_func=None):
-        # type: (Exception, int, bool, Any) -> None
+    def __init__(self, inner: Exception, delay: int = None, count_retries: bool = None, max_retries_func=None):
```
Comment (gurkansokat) [inline eb_sqs/auto_tasks/exceptions.py:2]: max_retries_func: Any=None?

**Comment by gurkansokat** [comment]
It seems like we are still expecting to pass circleci tests for python 2.7 and 3.7 still. Are those in project settings?

File: eb_sqs/auto_tasks/service.py [hunk 10]
```diff
@@ -24,7 +24,8 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
         class_ = getattr(module, class_name)  # find class
 
         auto_task_executor_service = _AutoTaskExecutorService(func_name)
-        instance = class_(auto_task_service=auto_task_executor_service)  # instantiate class using _AutoTaskExecutorService
+        instance = class_(
+            auto_task_service=auto_task_executor_service)  # instantiate class using _AutoTaskExecutorService
```
Comment (alexeyts) [inline eb_sqs/auto_tasks/service.py:28]: nit: indentation - closing bracket and comment could be on next line (and comment can probably be removed altogether)

just wanted to add a comment on this PR! 😁

Comment (gurkansokat) [reply eb_sqs/auto_tasks/service.py:28]: I thought GitHub was messing with me when I saw the name.. 😁 Good to see you watching...

File: eb_sqs/auto_tasks/service.py (hunk 3 above)
Comment (sherwinlu) [reply eb_sqs/auto_tasks/service.py:58]: i think we can just leave it as is. there's too many lines of code that use the old style and it's not worth the effort to convert imo - unless Python plans to not support that old style format anytime soon.

File: eb_sqs/auto_tasks/service.py [hunk 11]
```diff
@@ -82,14 +83,12 @@ def _auto_task_wrapper_invoker(*args, **kwargs):
...
-    def __init__(self, func_name):
-        # type: (str) -> None
+    def __init__(self, func_name: str):
         self._func_name = func_name
 
         self._executor_func_name = None
 
-    def register_task(self, method, queue_name=None, max_retries=None):
-        # type: (Any, str, int) -> None
+    def register_task(self, method, queue_name: str = None, max_retries: int = None):
```
[the remark this answers is not in the document]
Comment (sherwinlu) [reply eb_sqs/auto_tasks/service.py:91]: nope, was an oversight.

**Comment by sherwinlu** [comment]
@gurkansokat not sure why it's still looking for the python 2-7 and 3-7 tests. I removed them from the config.yml for circleci and replaced it with the python-3-9 test

File: eb_sqs/worker/service.py (hunk 7 above)
Comment (sherwinlu) [reply eb_sqs/worker/service.py:10]: yeah i'm getting the same errors too, but when i look at the documentation, it looks correct - https://boto3.amazonaws.com/v1/documentation/api/latest/guide/sqs.html#sqs

Comment (sherwinlu) [reply eb_sqs/worker/service.py]: actually when i looked again boto = boto 2.0 and we're using boto 3.0 so the reference is incorrect. however, i can't find Queue nor Message in the boto3 (and botocore) packages. So I've left it out for now.

File: eb_sqs/worker/worker_factory.py [hunk 12]
```diff
@@ -10,17 +10,14 @@ class WorkerFactory(object):
...
 
     @abstractmethod
-    def create(self):
-        # type: () -> Worker
+    def create(self) -> Worker:
         pass
 
     @staticmethod
     def default():
-        # type: () -> WorkerFactory
```
[the remark this answers is not in the document]
Comment (sherwinlu) [reply eb_sqs/worker/worker_factory.py:23]: It doesn't like it when I put WorkerFactory as the return type. It complains about not being able to find that class.

**Pull request merged**